src/programming-assistants.rst

```diff
@@ -1,23 +1,179 @@
+
+Gvero and Kuncak have a proposed a new code assistance tool, anyCode, for integrated development environments (IDEs) that produces a list of syntheses based on user-entered free-form queries :cite:`gvero2015synthesizing`. The queries contain a mixture of English and Java syntax and produce expressions that account for the correct syntax, types and scoping rules of Java.
+
+Given the query `open file example.txt`, anyCode would return a list of potential Java expressions:
+.. code-block:: Java
+  File file = new File("/Users/example/example.txt");
+  BufferedReader reader = new BufferedReader(new FileReader("/Users/example/example.txt"));
+  
+The authors state that the key feature that differentiates their tool is that it does not take into account any prior code context. The main take aways of their tool are that it often produces the expected code fragments, tolerates the flexibility of the English language, and it can repair incorrect Java expressions that use the wrong syntax or missing arguments.
+
+anyCode was evaluated with 45 examples where the effectiveness criteria was that it was able to successfully synthesize expected expressions and to list them among the top 10 solutions. The two models used to build the tool, unigram and PCFG, were evaluated separately and together. Using only the unigram model, the expected expression was only in the top 10 27% of the time and 18% using the PCFG model.
+
+However, when both models were used, the expected expression was in the top 10 82% of time and was the #1 option 44% of the time. More implementation details of the two models can be found `here <https://lara.epfl.ch/~kuncak/papers/GveroKuncak15SynthesizingJavaExpressionsFreeFormQueries.pdf>`_. anyCode was given an average runtime of 60 ms to generate the top 10 expressions, proving that it can effectively and efficiently synthesize expressions.
+
 Code Completion
 ---------------
 
+**Code Completion**, a subsection of Program Synthesis, refers to the process of intelligent completion suggestion making for lines of code a user writes during the development of an application. It is intended to speed up the process of coding applications by reducing typos and other mistakes. Code Completion assistants serves as convenient way to access function syntax and descriptions, significantly reducing the amount of memorization required. 
+
+Common techniques to achieve this include auto completion popups while typing, querying parameters of functions, querying hints related to syntax errors. The models used in Code Completion are very similar to those used in Program Synthesis but tend to use different inputs.
+
+Common inputs include:
+- Parsed code corpuses
```

Comment:
You should distinguish between training inputs, which come from code corpuses, from synthesis inputs, which come from the current editor state.